### ISSUE TITLE
Q: How records/coproducts with labels (rows?) normally formalized in staging?

### ISSUE THREAD

**Issue by anqur** [AndrasKovacs/staged#2]
Hi, Andras! It's me again watching this project for months and happy to see you're back! Here I have some little bit off-topic questions for you.

I'm really interested in ideas of staging and "program that generates programs" since my stargazing on this repo and I've been writing many code generation tools for my company products from last year (in poor Go, lol). But my scenerios are often like this:

> Reading **fields from structs** (labels from records) and generate other definitions, e.g. SQL.

So my question is that how fields or cases (or variants/branches? for enum/coproduct/sum types) or labels could be formalized to impose the staging gracefully? I mean in industry code there are many programs with structs/classes annotated with JSON/YAML/SQL information, which could definitely be computed during the meta theory. Would extension like row polymorphism be enough here?...

P.S. I couldn't entirely follow up all the details in your notes but I could roughly understand a little bit of some concepts like presheaf models, logical relations, etc. I would really love to hear your ideas about this as it might be a potential application of this project.

**Comment by AndrasKovacs** [comment]
I'm not completely sure what your question is about. If the goal is to convert some kind of untyped specification (JSON etc) at compile time to typed embedded syntax, that's fine, and how convenient it is really only depends on the expressivity of features that are available in the static language. The static language is compatible with a wide range of features, it may have any feature set that makes sense in a dependently typed language.

If you want row polymorphism or poly variants in the static language, that's completely fine. I guess what you want specifically is staged row/variant polymorphism, one without runutime overhead? That would look the pretty much the same as row/variant poly implementations in current languages, except that row constraints would live in the static universe.

**Comment by anqur** [comment]
> * If the goal is to convert some kind of untyped specification (JSON etc) at compile time to typed embedded syntax, that's fine, ...
> * If you want row polymorphism or poly variants in the static language, that's completely fine.

Yes! These are exactly the features I want and I could manage to think over. And...

> I'm not completely sure what your question is about. 

Pretty sorry for my description and let me rephrase it with a further example:

```java
public class Foo {
    @JSONField(name = "item-name") // Or `@ORMField(name = "item_name")`
    @Size(max = 10)
    String itemName;
}
```

Imagine that we have a record `Foo` that we want to:

* Handle the annoying casing options during JSON/ORM data convertion
* Use pre-defined sub-programs for data format validation like `@Size`

Intuitively if I'm the library writer that provides these annotations `@JSONField` and `@Size`, I need some meta-programming APIs like **getClassFields** to retrieve the information I need (e.g. the field's name and type, and those extra annotation stuff like `@Size` arguments and the names of `@JSONField`s).

And my further intuition is that, those annotations are simply static functions that would generate some intermediate records for carrying the necessary data, and then there would be another function to generate the *final* record type for runtime use. So my confusion is just around the `getClassFields` magic... I still have no idea of its formalization, is it a builtin `getClassFields : (T : U0) -> [String]` (or a list of dedicate records with richer information) good here?...

**Comment by AndrasKovacs** [comment]
These are in the area of "datatype-generic programming". The existing literature applies, and the only extra thing staging provides in this area is the guarantee that generic constructs don't persist to generated code. See e.g. GHC-generics, generics-SOP in Haskell, generic descriptions in Agda, generics with fixpoints of functors or indexed functors, and so on.

We have a type of type descriptions in the static language, ways to interpret type descriptions as runtime types, and a collection of "generic" functions which operate over all described runtime types, or certain subsets of them. We are allowed to do arbitrary dependently typed programming, and define descriptions of types as we want, and store whatever data we want inside descriptions. Some very primitive description of a record type would be

    data TyDesc : U1 = ...
    RecordDesc : U1 = List (TyDesc, String)
    ElTy : TyDesc -> ^U0 = ...
    ElRecordDesc : RecordDesc -> ^U0 = ...
    eq : (A : TyDesc) -> El A -> El A -> Bool = ...
    recordEq : (R : RecordDesc) -> ElRecordDesc R -> ElRecordDesc R -> Bool = ...

`eq` and `recordEq` are examples for generic functions.

The non-trivial part of generics is usually how to make the encoding overhead as small as possible, to make programming with descriptions similarly ergonomic as programming directly with native types. There is usually a mismatch between native types and type descriptions, because in a dep. language users can define whatever generic descriptions they want, but the set of built-in types and built-in features is more fixed.

Another possibility is to be able to look inside the structure of actual runtime types, and to "pattern matching" on them. This is not yet worked out in sufficient detail for 2LTT. It's called "intensional analysis" or "non-generative staging", and probably it only works for simple types in the runtime language, and doesn't work for polymorphic or dependent types there. But for ordinary programming purposes it could be still very useful. In this style, we can write

    isRecordTy : ^U0 -> Prop = ...
    recordEq : (A : ^U0) -> isRecordTy A -> A -> A -> Bool0

where `isRecordTy` is a predicate on runtime types that we can define by recursion + matching, and `recordEq` work similarly.

**Comment by anqur** [comment]
Wow ahhhhhh, thanks for your sincere reply and thorough explanation! It's a great load of materials for me to check out already, e.g the Haskell/Agda datatype-generic features/libraries/papers!

> It's called "intensional analysis" or "non-generative staging", ...

Yes I believe that much application code is so struggling on handwiring the *deep-copy*/*deep-equal*/*show* clutter and I just missed the fact that Rust (`#[derive(...)`]) or Haskell (`deriving`) have been actually doing these things with so much science. Great thanks again and closing this for now.
